Fixes #36834 - Add SecureBoot support for arbitrary operating systems to "Grub2 UEFI" PXE loaders

goarsna opened the pull request:
This feature consists of two patches, one for foreman and one for smart-proxy.

This patch introduces a new loader of kind `:PXEGrub2TargetOS` which allows to provide host-specific Network Bootstrap Programs (NPB) in order to enable network based installations for SecureBoot-enabled hosts.

SecureBoot expects to follow a chain of trust from the start of the host to the loading of Linux kernel modules. The very first shim that is loaded basically determines which distribution is allowed to be booted or kexec'ed until next reboot.

The existing "Grub2 UEFI SecureBoot" is not sufficient as it limits the possible installations to the vendor of the Foreman (Smart Proxy) host system.

Providing shim and GRUB2 by the vendor of the to-be-installed operating system allows Foreman to install any operating system on SecureBoot-enabled hosts over network.

To achieve this, the host's DHCP filename option is set to a shim path in a directory that is host-specific (contains MAC address). Corresponding shim and GRUB2 binaries are copied into that directory along with the generated GRUB2 configuration files as we know from "Grub2 UEFI".

The required binaries must be provided once in the so called "bootloader universe". This directory can be configured via the settings file `/etc/foreman-proxy/settings.d/tftp.yml` and defaults to `/usr/local/share/bootloader-universe/<os>/`. These binaries can be manually retrieved from the installation media and is not part of this patch set.

Full example:
-------------

[root@vm ~]# hammer host info --id 241 | grep -E "(MAC address|Operating System)"
    MAC address:  00:50:56:b4:75:5e
    Operating System:       Ubuntu 22.04 LTS

[root@vm ~]# tree /usr/local/share/bootloader-universe/ /usr/local/share/bootloader-universe/
|-- centos
|   |-- grubx64.efi
|   `-- shimx64.efi
`-- ubuntu
    |-- grubx64.efi
    `-- shimx64.efi

[root@vm ~]# hammer host update --id 241 --build true

[root@vm ~]# tree /var/lib/tftpboot/grub2/00-50-56-b4-75-5e/ /var/lib/tftpboot/grub2/00-50-56-b4-75-5e/
|-- grub.cfg
|-- grub.cfg-00:50:56:b4:75:5e
|-- grub.cfg-01-00-50-56-b4-75-5e
|-- grubx64.efi
|-- shimx64.efi
`-- targetos

[root@vm ~]# grep -B2 00-50-56-b4-75-5e /var/lib/dhcpd/dhcpd.leases
  hardware ethernet 00:50:56:b4:75:5e;
  fixed-address 192.168.145.84;
        supersede server.filename = "grub2/00-50-56-b4-75-5e/shimx64.efi";

[root@vm ~]# pesign -S -i /var/lib/tftpboot/grub2/00-50-56-b4-75-5e/grubx64.efi | grep Canonical The signer's common name is Canonical Ltd. Secure Boot Signing (2021 v1)

jloeser committed:
Fixes #36834 - Add SecureBoot support for arbitrary operating systems to "Grub2 UEFI" PXE loaders

This feature consists of four patches, one each for foreman,
smart-proxy, foreman-installer, and puppet-foreman_proxy.

This patch adds support for individual Network Bootstrap Programs (NBP)
in order to enable network based installations of SecureBoot enabled
hosts for arbitrary operating systems.

SecureBoot expects to follow a chain of trust from the initial boot of
the host to the loading of Linux kernel modules. The very first shim
that is loaded determines which distribution is allowed to be booted or
kexec'ed until next reboot.

Currently the "Grub2 UEFI SecureBoot" PXE loaders use NBPs provided by
the vendor of the Foreman/Smart Proxy host system. All hosts receive and
execute the same binary. On SecureBoot enabled hosts, this limits
installations to operating systems by the vendor of the Foreman/
Smart Proxy host system.

Providing shim and GRUB2 by the vendor of the operating system to be
installed allows Foreman to install any operating system on SecureBoot
enabled hosts over network.

To achieve this, the host's DHCP filename option is set to a shim/GRUB2
binary in a host specific directory based on their MAC address.
Corresponding shim and GRUB2 binaries are copied into that directory
along with the generated GRUB2 configuration files.
When provisioning a host, the Smart Proxy checks in a dedicated
directory inside the TFTP root - the so called "bootloader universe" -
if NBPs are present matching the operating system, operating system
version, and architecture of the host to be installed. If this is the
case, these NBPs are copied from the bootloader universe directory to
the host specific directory. If not, as a fallback the default NBPs
provided by the vendor of the Foreman/Smart Proxy host system are
copied from the `:tftproot:/grub2` directory to the host specific
directory.

Up to now, shim and GRUB2 binaries have to be retrieved and set up in
the bootloader universe directory manually according to the
documentation. An automatic way to provide OS dependent NBPs will be
added in future.

In case there are no NBPs present in the bootloader universe matching
the operating system, operating system version, and architecture of
the host to be installed, the behaviour of the "Grub2 UEFI" PXE
loaders does not change to the behavior prior to this feature.

Implementation notes:
---------------------
* To be future proof (e.g. to be able to provide NBPs in the bootloader
  universe for other PXE loaders without running into any filename
  conflicts) and for better structure, the PXE kind is prepended as a
  first directory level inside the bootloader universe.
* The operating system version inside the bootloader universe consists
  of the major and minor version (if applicable) of the operating system
  separated by a dot (`.`). If no NBPs are configured for a specific
  operating system version the fallback directory `default` is used.
* To simplify things on Foreman side in future, symlinks are used for
  the shim (boot-sb.efi) and GRUB2 (boot.efi) binaries.
* Inside the TFTP root directory a new directory `host-config` is
  created for storing all the host specific directories.
* Inside the TFTP root directory a new directory `bootloader-universe`
  is created for storing all the OS specific boot files.
* For storage efficiency the shim and GRUB2 binaries from the
  bootloader universe or the `:tftproot:/grub2` directory are
  symlinked to the host specific directory.

Full example:
-------------
[root@vm ~]# hammer host info --id 241 | grep -E "(MAC address|Operating System)"
MAC address: 00:50:56:b4:75:5e
Operating System: AlmaLinux 8.9

[root@vm ~]# tree /var/lib/tftpboot/bootloader-universe/
/var/lib/tftpboot/bootloader-universe/
└── pxegrub2
    └── almalinux
        ├── 8.9
        │   └── x86_64
        │       ├── boot.efi -> grubx64.efi
        │       ├── boot-sb.efi -> shimx64.efi
        │       ├── grubx64.efi
        │       └── shimx64.efi
        └── default
            └── x86_64
                ├── boot.efi -> grubx64.efi
                ├── boot-sb.efi -> shimx64.efi
                ├── grubx64.efi
                └── shimx64.efi

[root@vm ~]# hammer host update --id 241 --build true

[root@vm ~]# tree /var/lib/tftpboot/host-config
/var/lib/tftpboot/host-config
└── 00-50-56-a3-41-a8
    └── grub2
        ├── boot.efi -> ../../../bootloader-universe/grubx64.efi
        ├── boot-sb.efi -> ../../../bootloader-universe/shimx64.efi
        ├── grub.cfg
        ├── grub.cfg-00:50:56:a3:41:a8
        ├── grub.cfg-01-00-50-56-a3-41-a8
        ├── grubx64.efi -> ../../../bootloader-universe/grubx64.efi
        ├── os_info
        └── shimx64.efi -> ../../../bootloader-universe/shimx64.efi

[root@vm ~]# grep -B2 00-50-56-b4-75-5e /var/lib/dhcpd/dhcpd.leases
hardware ethernet 00:50:56:b4:75:5e;
fixed-address 192.168.145.84;
supersede server.filename = "host-config/00-50-56-b4-75-5e/grub2/boot-sb.efi";

[root@vm ~]# pesign -S -i /var/lib/tftpboot/host-config/00-50-56-b4-75-5e/grub2/boot-sb.efi | grep "Microsoft Windows UEFI Driver Publisher"
The signer's common name is Microsoft Windows UEFI Driver Publisher

goarsna commented:
I have opened a PR to document this feature in foreman-documentation:
https://github.com/theforeman/foreman-documentation/pull/2145

goarsna commented:
This PR also requires the changes from Foreman Smart Proxy PR [#877](https://github.com/theforeman/smart-proxy/pull/877)

ekohl commented on app/models/concerns/pxe_loader_support.rb:
```diff
@@ -28,6 +29,7 @@ def all_loaders_map(precision = 'x64', httpboot_host = "httpboot_host")
         "Grub2 ELF" => "grub2/grub#{precision}.elf",
         "Grub2 UEFI" => "grub2/grub#{precision}.efi",
         "Grub2 UEFI SecureBoot" => "grub2/shim#{precision}.efi",
+        "Grub2 UEFI SecureBoot (target OS)" => "grub2/@@subdir@@/shim#{precision}.efi",
```
I wonder if we can somehow guard this on the capability I spoke about in https://github.com/theforeman/smart-proxy/pull/877#pullrequestreview-1689754274 so the user can only select it if it's supported.

goarsna replied:
Thanks @ekohl, this would be great! I will read into that topic and have a look how I can guard this.

goarsna replied:
Hey @ekohl, I had a deeper look at the changes that would be necessary to add the new PXE loader to the list of PXE loaders depending on the smart proxy capability. According to my findings, this would require changes in more than 15 additional files.
In the Web UI the list of PXE loaders is loaded when an operating system is selected in the Host Edit page. As the Smart Proxy depends on the Subnet, we would have to pass the currently selected Subnet down from that point through [operatingsystem.rb](https://github.com/theforeman/foreman/blob/develop/app/models/operatingsystem.rb) (and all derived classes) to the point where the dictionary of PXE loaders is assembled in [pxe_loader_support.rb](https://github.com/theforeman/foreman/blob/develop/app/models/concerns/pxe_loader_support.rb). 
Furthermore we would also have to guard the available PXE loaders in the parameter validation in the [hostgroups_controller.rb](https://github.com/theforeman/foreman/blob/develop/app/controllers/api/v2/hostgroups_controller.rb) and [hosts_controller.rb](https://github.com/theforeman/foreman/blob/develop/app/controllers/api/v2/hosts_controller.rb) (both for parameter validations). And there are still some more references to check for required changes.
We would also have to reload the PXE loaders in case the Subnet (and therefore the Smart Proxy) is changed and handle cases where the new PXE loader isn’t available any more after the change of the Subnet. I have not identified yet what would be necessary for that. But this also means, that we cannot freeze the dictionary of PXE loaders for each operating system upon creation anymore, as we have to regenerate it on every Subnet change.

All in all this seems a lot of effort for guarding the list of available PXE loaders depending on the Smart Proxy capability. So I looked for an alternative solution:

I now was thinking about guarding the new PXE loader at the moment the changes are submitted in the Web UI and in the actual handling of the API requests. What do you think?

goarsna replied:
The behavior has completely changed with our new approach. The Smart Proxies capabilities will be checked in module `Orchestration::DHCP` when the boot file name is set in the DHCP attributes. If a Smart Proxy does not provide the required capability the legacy boot file names will be used.

sbernhard commented on app/models/concerns/orchestration/tftp.rb:
```diff
@@ -82,12 +82,20 @@ def default_pxe_render(kind)
+      content = generate_pxe_template("PXEGrub2".to_sym)
+    else
+      content = generate_pxe_template(kind)
+    end
```
```
    content = if kind.to_s == "PXEGrub2TargetOS"
                generate_pxe_template("PXEGrub2".to_sym)
              else
                generate_pxe_template(kind)
              end
  ```

sbernhard commented on app/models/concerns/orchestration/tftp.rb:
```diff
@@ -82,12 +82,20 @@ def default_pxe_render(kind)
       each_unique_feasible_tftp_proxy do |proxy|
         mac_addresses_for_provisioning.each do |mac_addr|
-          proxy.set(kind, mac_addr, :pxeconfig => content)
+          targetos = nil
```
```
          targetos = if kind.to_s == "PXEGrub2TargetOS"
            host.operatingsystem.name.downcase
          else
            nil
          end
```

sbernhard commented on app/models/concerns/orchestration/tftp.rb:
```diff
@@ -82,12 +82,20 @@ def default_pxe_render(kind)
         mac_addresses_for_provisioning.each do |mac_addr|
-          proxy.set(kind, mac_addr, :pxeconfig => content)
+          targetos = nil
+          if kind.to_s == "PXEGrub2TargetOS"
```
I think it would be good to have a helper method like
```
def isPXEGrub2TargetOS
  return true if kind.to_s == "PXEGrub2TargetOS"
  false
end
```

sbernhard commented on app/models/concerns/orchestration/dhcp.rb:
```diff
@@ -125,6 +125,10 @@ def dhcp_attrs(record_mac)
       filename = operatingsystem.boot_filename(host)
+      if filename.include? "@@subdir@@"
+        mac = host.mac.downcase
+        filename = filename.gsub("@@subdir@@", mac.tr(':', '-').downcase)
```
2 times downcase?

goarsna replied:
Oh, thanks.. I reworked that part

sbernhard commented on app/models/concerns/orchestration/tftp.rb:
```diff
@@ -132,6 +140,9 @@ def validate_tftp
     return unless host.operatingsystem
     pxe_kind = host.operatingsystem.pxe_loader_kind(host)
+    if pxe_kind.to_s == "PXEGrub2TargetOS"
+      pxe_kind = "PXEGrub2".to_sym
```
I would suggest to have a mapping function somehwere so that no hard coding is necessary here,
Like
```
pxe_kind_mapping = {
  'PXEGrub2TargetOS' => 'PXEGrub'
}
```
What do you think about this @ekohl - Maybe you have an good idea where this mapping should be.

goarsna replied:
Deprecated with our new approach

sbernhard commented on app/models/concerns/orchestration/tftp.rb:
```diff
@@ -146,7 +157,11 @@ def queue_tftp
   def queue_tftp_create
+    pxe_kind = host.operatingsystem.pxe_loader_kind(host)
     host.operatingsystem.template_kinds.each do |kind|
+      if kind.to_s == "PXEGrub2" && pxe_kind.to_s == "PXEGrub2TargetOS"
```
Do you have an idea how to implement this in a more generic way? Or should we really have hard-coded this.

goarsna replied:
Deprecated with our new approach

sbernhard commented on app/models/concerns/orchestration/tftp.rb:
```diff
@@ -132,6 +137,9 @@ def validate_tftp
     return unless host.operatingsystem
     pxe_kind = host.operatingsystem.pxe_loader_kind(host)
+    if isPXEGrub2TargetOS(pxe_kind)
+      pxe_kind = "PXEGrub2".to_sym
```
pxe_kind = "PXEGrub2".to_sym if isPXEGrub2TargetOS(pxe_kind)

goarsna commented:
Hi there,
I thought a lot about how to simplify things in the `Orchestration::TFTP` module. I introduced a [PXE_KIND_MAP](https://github.com/theforeman/foreman/pull/9864/commits/f1dd4c321e6778bc0dbb6eeeb52258ee1e1cd1e7#diff-675db90052a00f189a0cd62d48f5adcef7b9a7ed6c3d2b846869260a665df71c) in `pxe_loader_support.rb` to map template kinds to the respective template kind that should be used for rendering. By this the changes in [tftp.rb](https://github.com/theforeman/foreman/pull/9864/commits/f1dd4c321e6778bc0dbb6eeeb52258ee1e1cd1e7#diff-0c5daa2f2b58e50f465a3a566a7a46d2994a90fa4358e43721155e4f02ab6a04) become smaller and simpler.

The other thing I was investigating is why the `queue_tftp_create` method in [tftp.rb](https://github.com/theforeman/foreman/pull/9864/commits/fdc23d75a45e669686949cdb0fb24f53112a3ef5#diff-0c5daa2f2b58e50f465a3a566a7a46d2994a90fa4358e43721155e4f02ab6a04) loops over all template kinds the chosen operating system supports? As I couldn't find a good reason for it after all investigations I did, I changed the implementation. Now only the template for the chosen template kind gets rendered, what allowed to drop the `if`s and `else`s and makes this method also pretty short.

Any thoughts or objections?

goarsna commented:
Hey there,
unfortunately the PXE loader mapping wasn't working. I fixed it now.
Also templates weren´t generated in case the adjusted template kind is changed. This is necessary if we omit the loop, as only the selected template kind gets rendered.

What do you think, @ekohl?

And I just saw, that there are some failing tests. :sweat_smile: I'll do that ASAP.

goarsna commented:
Hey @ekohl,
I had a lot of discussions with people here at ATIX. And we figured out that the best approach is to keep the loop in `queue_tftp_create` and instead add the new PXE loader to the default `template_kinds` in `operatingsystem.rb`. By this everything works as expected and the introduced changes are minimized.

But we definitely require some feedback, weather this PR is going in the right direction and regarding the Smart Proxy capability (see my [question above](https://github.com/theforeman/foreman/pull/9864#discussion_r1401116524)).

Thanks!

ekohl commented on app/models/concerns/orchestration/tftp.rb:
```diff
@@ -82,12 +82,12 @@ def default_pxe_render(kind)
       each_unique_feasible_tftp_proxy do |proxy|
         mac_addresses_for_provisioning.each do |mac_addr|
-          proxy.set(kind, mac_addr, :pxeconfig => content)
+          proxy.set(kind, mac_addr, {:pxeconfig => content, :targetos => host.operatingsystem.name.downcase})
```
What would this do on a Smart Proxy that doesn't support this feature? We have support for using Foreman X.Y with Smart Proxy X.(Y-1) in Satellite (to make upgrading easier for large deployments) so it's always useful to keep that in mind.

goarsna replied:
As we are only adding post parameters here, in case a "old" Smart proxy is used, they are simply ignored.

goarsna commented:
Hi there,
I have rethought the whole concept of how we wanted to add SecureBoot support for arbitrary operating systems due to the discussions in [Foreman Community Forum](https://community.theforeman.org/t/add-secureboot-support-for-arbitrary-distributions/32601/21) and came up with an idea how to completely integrate what we want into the existing `PXEGrub2` PXE loaders.

What's left to do for now:
* I will have to take a look at the tests after my changes.
* As this approach generates MAC specific directories directly inside `/var/lib/tftpboot/` the ownership of this directory has to be changed to user `foreman-proxy`.
* One open request by @lzap was to provide differentiated bootloader files in the `bootloader_universe` not only by OS vendor but also by OS version.

@sbernhard @ekohl What do you think? Please share your thoughts. I'm looking forward to input and discussion.

goarsna commented:
Implemented changes as discussed in the course of the [Foreman Community Forum](https://community.theforeman.org/t/add-secureboot-support-for-arbitrary-distributions/32601/21) thread.

Please let me know, if anything is missing or further changes are required.

goarsna commented:
Added deletion of host specific bootloader files in case host leaves build mode.

goarsna commented on app/models/concerns/pxe_loader_support.rb:
```diff
@@ -5,7 +5,7 @@ module PxeLoaderSupport
   PXE_KINDS = {
     :PXELinux => /^(pxelinux.*|PXELinux (BIOS|UEFI))$/,
     :PXEGrub => /^(grub\/|Grub UEFI).*/,
-    :PXEGrub2 => /^(grub2\/|Grub2 (BIOS|UEFI|ELF)|http.*grub2\/).*/,
```
FYI: While fixing the tests I stumbled across this mapping of filename / loader name to template kind, because a test was broken due to the changes of the filenames down below. I found out that actually a mapping from filename to template kind is never used in the code (besides in the tests). There is exactly one usage of this mapping and there only the loader name is used. So I was wondering if there is a historic reason for the mapping by filename? (cc @lzap)

lzap replied:
If `git log -Gxxxx` does not explain it, then delete, yeah. I don’t remember to be honest.

goarsna replied:
Ok, searching with `git log` reveals more places I would want to check before removing the filename mapping than I can cope with for now. Better to do this in a separate PR.

goarsna commented:
Still some failing tests, but most of them I was able to fix up to now.

jloeser commented on app/models/concerns/orchestration/tftp.rb:
```diff
@@ -87,7 +87,14 @@ def setTFTP(kind)
+                      :targetos => host.operatingsystem.name.downcase,
+                      :major => host.operatingsystem.major,
+                      :minor => host.operatingsystem.minor,
+                      :arch => host.arch.bootfilename_efi,
```
Let's use `host.arch.name` here which is at the end the architecture string reported by the system with `uname -m` (e.g. `x86_64`).

goarsna replied:
Agreed, it's better to use `host.arch.name` here. => Changed and tested locally.

goarsna commented:
Squashed and changed commit message to reflect current implementation. => Any objections against merging this? What do you think, @nofaralfasi and @stejskalleos?

stejskalleos commented:
Sorry for the delay, this week I'll take a look at it for sure!

nofaralfasi commented on test/models/concerns/pxe_loader_support_test.rb:
```diff
@@ -32,18 +32,28 @@ def setup
-    test "PXEGrub2 is found for given filename" do
-      @host.pxe_loader = "grub2/grubx64.efi"
+    test "PXEGrub2 is found for grubx64.elf filename" do
+      @host.pxe_loader = "grub2/grubx64.elf"
```
What is `grubx64.elf`? Shouldn't it be `grubx64.efi`?

goarsna commented:
Given that I only added two lines of comments and that the tests were successful before, I assume the failing tests are not related to my changes.

stejskalleos commented on app/models/concerns/orchestration/dhcp.rb:
```diff
@@ -125,6 +125,15 @@ def dhcp_attrs(record_mac)
     if provision?
       dhcp_attr[:nextServer] = boot_server unless host.pxe_loader == 'None'
       filename = operatingsystem.boot_filename(host)
+      if filename.include? "@@subdir@@"
```
Can we have it as a separate method? The `dhcp_attrs` is already long enough

stejskalleos commented on app/models/concerns/orchestration/dhcp.rb:
```diff
@@ -125,6 +125,15 @@ def dhcp_attrs(record_mac)
       dhcp_attr[:nextServer] = boot_server unless host.pxe_loader == 'None'
       filename = operatingsystem.boot_filename(host)
+      if filename.include? "@@subdir@@"
+        if host.subnet.tftp.has_capability?(:TFTP, :target_os_bootloader_support)
```
Can the `host.subnet.tftp` be `nil`? If it can happen, this will fail with the `undefined method has_capability? for nil:NilClass` error

stejskalleos replied:
Note to me: What will happen if the host's subnet doesn't have any tftp?

goarsna replied:
Good call :sweat_smile: I always thought a subnet without TFTP doesn't make any sense for provisioning and missed that scenario. But yes, in case subnet or tftp is nil we don't want to see an undefined method error...

goarsna commented:
Thanks @stejskalleos for your review. My last commit is simply a rebase.

goarsna commented:
See [discussion in smart-proxy PR](https://github.com/theforeman/smart-proxy/pull/877#discussion_r1657078777) for information on recent changes.

goarsna commented:
Rebased to head of develop

goarsna commented:
I renamed the `host-config` directory as decided in the course of the [above mentioned discussion](https://github.com/theforeman/foreman/pull/9864#issuecomment-2207054058). Besides that I also cleaned up some changes that have already been obsoleted by prior discussions / comments in this PR.

goarsna commented:
See [this comment](https://github.com/theforeman/smart-proxy/pull/877#discussion_r1668840574) for the latest update.

goarsna commented:
While reviewing the changes again, I discovered further obsoleted changes in the `tftp_test.rb`.

goarsna commented:
See this [comment in the Smart Proxy PR](https://github.com/theforeman/smart-proxy/pull/877?notification_referrer_id=NT_kwDOAkJiPbM4MDY3MzgzMjE5OjM3OTA0OTU3#discussion_r1761516220) regarding the recent changes.

nofaralfasi commented:
/packit build

nofaralfasi commented:
Blocked by: https://github.com/theforeman/foreman-installer/pull/949/

stejskalleos commented:
Thanks, everyone, for the effort and testing, great work!

goarsna commented:
Also from my side a big thank you to everyone who was involved in getting this merged :star_struck:
